xio-core/src/test/java/com/xjeffrose/xio/http/Http2To1ProxyRequestQueueTest.java

```diff
@@ -0,0 +1,178 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import com.xjeffrose.xio.http.internal.Http1SegmentedData;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.DefaultLastHttpContent;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class Http2To1ProxyRequestQueueTest extends Assert {
+
+  @Mock ChannelHandlerContext mockCtx;
+  private Http2To1ProxyRequestQueue subject;
+
+  @Before
+  public void beforeEach() {
+    MockitoAnnotations.initMocks(this);
+    subject = new Http2To1ProxyRequestQueue();
+  }
+
+  @Test
+  public void testInterleavedH2FrontRequests_stateTransitions() {
+    // should start empty
+    assertTrue(subject.isEmpty());
+
+    // the first stream id should write
+    subject.onRequestWriteOrEnqueue(mockCtx, 3, "request 1a", mock(ChannelPromise.class));
+    verify(mockCtx).write(eq("request 1a"), any());
+    assertTrue(subject.isEmpty());
+
+    // the first stream id should write again
+    subject.onRequestWriteOrEnqueue(mockCtx, 3, "request 1b", mock(ChannelPromise.class));
+    verify(mockCtx).write(eq("request 1b"), any());
+    assertTrue(subject.isEmpty());
+
+    // subsequent stream ids should be enqueued
+    subject.onRequestWriteOrEnqueue(mockCtx, 5, "request 2a", mock(ChannelPromise.class));
+    verify(mockCtx, never()).write(eq("request 2a"), any());
+    assertFalse(subject.isEmpty());
+
+    // the first stream id should write again
+    subject.onRequestWriteOrEnqueue(mockCtx, 3, "request 1c", mock(ChannelPromise.class));
+    verify(mockCtx).write(eq("request 1c"), any());
+    assertFalse(subject.isEmpty());
+
+    // when the first stream id response completes
+    assertEquals(3, subject.currentProxiedH2StreamId().orElse(0), 0);
+    Response response =
+        DefaultFullResponse.builder()
+            .headers(new DefaultHeaders())
+            .status(HttpResponseStatus.OK)
+            .body(Unpooled.EMPTY_BUFFER)
+            .streamId(3)
+            .build();
+    subject.onResponseDrainNext(mockCtx, response);
+
+    // then the subsequent stream ids should write
+    verify(mockCtx).writeAndFlush(eq("request 2a"), any());
+    assertSame(response, subject.currentResponse().orElse(null));
+
+    // and the stream id should be correct
+    assertEquals(5, subject.currentProxiedH2StreamId().orElse(0), 0);
+
+    // and the the queue should be empty
+    assertTrue(subject.isEmpty());
+  }
+
+  @Test
+  public void testInterleavedH2FrontRequestsSegmentedResponse_stateTransitions() {
+    // should start empty
```

Comment:
yeah this test is too large - i'm not sure how else to cover all the state transitions